jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java

```diff
@@ -0,0 +1,343 @@
+                newLabeledPath.labels = new HashSet<>(this.labels);
+                newLabeledPath.cost = this.cost;
+
+                return newLabeledPath;
+            } catch (CloneNotSupportedException e) {
+                e.printStackTrace();
+                throw new RuntimeException();
+            }
+        }
+    }
+
+    private Graph<V, E> graph;
+    private Map<V, Integer> labels;
+    private int lengthBound;
+
+    /**
+     * Constructs an algorithm with given inputs
+     *
+     * @param graph the (improvement) graph on which to calculate the local augmentation
+     * @param lengthBound the upper bound for the length of cycles to detect
+     * @param labels the labels of the vertices encoding the subsets of vertices
+     */
+    public AhujaOrlinSharmaCyclicExchangeLocalAugmentation(Graph<V, E> graph, int lengthBound, Map<V, Integer> labels) {
+        this.graph = Objects.requireNonNull(graph, "Graph cannot be null");
+        this.lengthBound = lengthBound;
+        this.labels = Objects.requireNonNull(labels, "Labels cannot be null");
+    }
+
+    /**
+     * Calculates a valid subset-disjoint negative cycle.
+     * If there is no such cycle, it returns an empty LabeledPath instance with cost Double.MAX_VALUE.
+     *
+     * @return a valid subset-disjoint negative cycle encoded as LabeledPath
+     */
+    public GraphWalk<V, E> getLocalAugmentationCycle() {
+
+        int k = 1;
+        LabeledPath<V> cycleResult = new LabeledPath<>(lengthBound);
+
+        /*
+         * Store the path sets via a with Pair<V, V> as for the head and tail of the path such that the domination tests runs faster.
+         * Furthermore, use a TreeSet to sort for the cost of the paths such that the domination test stops optimally concerning the cost of the paths.
+         */
+        Map<Pair<V, V>, TreeSet<LabeledPath<V>>> pathsLengthK = new LinkedHashMap<>();
+        Map<Pair<V, V>, TreeSet<LabeledPath<V>>> pathsLengthKplus1 = new LinkedHashMap<>();
+
+        // initialize pathsLengthK for k = 1
+        for(E e : graph.edgeSet()) {
+            if(graph.getEdgeWeight(e) < 0) {
+                // initialize all paths of cost < 0
+                ArrayList<V> pathVertices = new ArrayList<>(lengthBound);
+                Set<Integer> pathLabels = new HashSet<>();
+                V sourceVertex = graph.getEdgeSource(e);
+                V targetVertex = graph.getEdgeTarget(e);
+                pathVertices.add(sourceVertex);
+                pathVertices.add(targetVertex);
+                pathLabels.add(labels.get(sourceVertex));
+                pathLabels.add(labels.get(targetVertex));
+                LabeledPath<V> path = new LabeledPath<>(pathVertices, graph.getEdgeWeight(e), pathLabels);
+
+                // add path to set of paths of length 1
+                addPathToPathStructure(pathsLengthK, path);
+            }
+        }
+
+        while (k < lengthBound && cycleResult.getCost() >= 0) {
```

Comment:
Won't cycleResult always be nonnegative here since if we hit a negative cycle inside of this loop, we return immediately?  In fact, I can't understand why cycleResult needs to be maintained across loops at all.  It should only be constructed directly at the point of return unless it's being used for some kind of minimization/maximization.